### PR TITLE
add `App.validators`, staking module, Validator type and track initial validator set

### LINKED ISSUE #36: Track validators in ABCI application state
To do staking and delegation, the Penumbra application needs to track the validator set and maintain records of who the validators are.

### PR DESCRIPTION
**Pull request by avahowell**
This PR adds the `validators` field to `App`, which is a `BTreeMap<tendermint::PublicKey, Validator>` that will track the validator set. Note that this just pulls the initial validator state from the tendermint genesis state, we will need to add additional logic to the ABCI app once undelegations, slashing/evidence handling, etc, are implemented.

Closes #36

### CODE REVIEW

File: pd/src/staking.rs
Comment (hdevalence) [inline pd/src/staking.rs:9]: non-blocking on this PR, but in the future we might want to merge this into types in https://rustdoc.penumbra.zone/main/tendermint/validator/index.html (and/or modify them to suit our purposes; the `tendermint-rs` API isn't totally fleshed out yet)

File: pd/src/app.rs
Comment (hdevalence) [inline pd/src/app.rs:81]: It looks like at the moment we don't need to share access to the `validators`, since it's set in the body of `init_genesis` where there's already a `&mut self` available. Are we expecting to need to read or write it in other places where we have to do asynchronous processing?

**Comment by hdevalence** [comment]
Do we need to save the validator info to the application state?

**Comment by avahowell** [comment]
Yes, I think so, working on that now

File: pd/src/staking.rs
Comment (avahowell) [reply pd/src/staking.rs:9]: I wasn't sure if we wanted to have a separate type specifically for Penumbra validators: right now this type is the same as the Tendermint type, but in the future I was imagining that we may want to have more penumbra-specific data or methods in this struct

File: pd/src/app.rs
Comment (avahowell) [reply pd/src/app.rs:81]: I think we'll need to write to this in `end_block` (for validator updates) and `begin_block` (for handling evidence and rewards)

File: pd/src/staking.rs
Comment (hdevalence) [reply pd/src/staking.rs:9]: Yeah, starting with our own type is a good call, just wanted to make the connection for the future.

File: pd/src/app.rs
Comment (hdevalence) [reply pd/src/app.rs:81]: Also non-blocking for this PR, but just a comment for the future: we might want to explore splitting out these fields into an `Inner` type, so that rather than having multiple chunks of synchronization, we can wrap all of the state that we need to access in one mutex.

**Comment by avahowell** [comment]
added code that saves validators to the application state on init genesis and retrieves them in the constructor of `App`

File: pd/src/staking.rs
Comment (hdevalence) [inline pd/src/staking.rs:3]: non-blocking, but I think cosmos uses `cosmosvaloper` or something, maybe we should match it exactly?

File: pd/src/state.rs
Comment (hdevalence) [inline pd/src/state.rs:269]: non-blocking on this PR: we have some centralization of our database writes into the `commit` method, maybe it would make sense to merge this setup code with the setup code in #234 into a single method that would be more clearly a call-once method?

**Review by hdevalence — APPROVED** [review]
This LGTM, only comments are non-blocking. Currently there's a merge conflict so I'll rebase and de-conflict before merging.

**Pull request merged**
